File: routes/empresas.py
from flask import Blueprint
from flask import render_template
from flask import request
from flask import redirect
from flask import session
from flask import jsonify

from services.empresa_service import EmpresaService

empresas_bp = Blueprint("empresas", __name__)


# ==========================
# LISTAR EMPRESAS
# ==========================

@empresas_bp.route("/empresas")
def empresas():

    if "usuario" not in session:
        return redirect("/")

    lista = EmpresaService.listar()

    return render_template(
        "empresas.html",
        usuario=session["usuario"],
        perfil=session["perfil"],
        empresas=lista
    )


# ==========================
# NOVA EMPRESA
# ==========================

@empresas_bp.route("/empresas/nova", methods=["POST"])
def nova_empresa():

    if "usuario" not in session:
        return jsonify({
            "sucesso": False,
            "mensagem": "Sessão expirada."
        }), 401

    dados = {

        "razao_social": request.form.get("razao_social", "").strip(),
        "nome_fantasia": request.form.get("nome_fantasia", "").strip(),
        "cnpj": request.form.get("cnpj", "").strip(),
        "inscricao_estadual": request.form.get("inscricao_estadual", "").strip(),
        "responsavel": request.form.get("responsavel", "").strip(),
        "telefone": request.form.get("telefone", "").strip(),
        "whatsapp": request.form.get("whatsapp", "").strip(),
        "email": request.form.get("email", "").strip(),
        "cep": request.form.get("cep", "").strip(),
        "endereco": request.form.get("endereco", "").strip(),
        "numero": request.form.get("numero", "").strip(),
        "bairro": request.form.get("bairro", "").strip(),
        "cidade": request.form.get("cidade", "").strip(),
        "estado": request.form.get("estado", "").strip(),
        "status": request.form.get("status", 1)

    }

    if EmpresaService.cnpj_existe(dados["cnpj"]):

        return jsonify({
            "sucesso": False,
            "mensagem": "Já existe uma empresa cadastrada com este CNPJ."
        })

    EmpresaService.inserir(dados)

    return jsonify({
        "sucesso": True,
        "mensagem": "Empresa cadastrada com sucesso."
    })


# ==========================
# EDITAR EMPRESA
# ==========================

@empresas_bp.route("/empresas/editar/<int:id>", methods=["POST"])
def editar_empresa(id):

    if "usuario" not in session:
        return redirect("/")

    dados = {

        "razao_social": request.form.get("razao_social", ""),
        "nome_fantasia": request.form.get("nome_fantasia", ""),
        "cnpj": request.form.get("cnpj", ""),
        "inscricao_estadual": request.form.get("inscricao_estadual", ""),
        "responsavel": request.form.get("responsavel", ""),
        "telefone": request.form.get("telefone", ""),
        "whatsapp": request.form.get("whatsapp", ""),
        "email": request.form.get("email", ""),
        "cep": request.form.get("cep", ""),
        "endereco": request.form.get("endereco", ""),
        "numero": request.form.get("numero", ""),
        "bairro": request.form.get("bairro", ""),
        "cidade": request.form.get("cidade", ""),
        "estado": request.form.get("estado", ""),
        "status": request.form.get("status", 1)

    }

    EmpresaService.atualizar(id, dados)

    return redirect("/empresas")


# ==========================
# EXCLUIR EMPRESA
# ==========================

@empresas_bp.route("/empresas/excluir/<int:id>", methods=["POST"])
def excluir_empresa(id):

    if "usuario" not in session:
        return redirect("/")

    EmpresaService.excluir(id)

    return redirect("/empresas")

File: static/js/empresas.js
// =========================
// ELEMENTOS DA PÁGINA
// =========================

const modalEmpresa = document.getElementById("modalEmpresa");
const modalExcluir = document.getElementById("modalExcluir");

const empresaExcluir = document.getElementById("empresaExcluir");
const formExcluir = document.getElementById("formExcluir");
const formEmpresa = document.getElementById("formEmpresa");


// =========================
// EXCLUIR EMPRESA
// =========================

document.querySelectorAll("a.excluir").forEach(function (botao) {

    botao.addEventListener("click", function (e) {

        e.preventDefault();

        empresaExcluir.innerText = this.dataset.razao;

        formExcluir.action =
            "/empresas/excluir/" + this.dataset.id;

        modalExcluir.classList.add("active");

    });

});

function fecharModalExcluir() {

    modalExcluir.classList.remove("active");

}


// =========================
// ABRIR / FECHAR MODAL
// =========================

function abrirModal() {

    modalEmpresa.classList.add("active");

    formEmpresa.reset();

    document.getElementById("empresa_id").value = "";

    formEmpresa.action = "/empresas/nova";

    limparMensagem();

}

function fecharModal() {

    modalEmpresa.classList.remove("active");

}


// =========================
// MENSAGENS
// =========================

function obterMensagem() {

    let msg = document.getElementById("mensagemErro");

    if (!msg) {

        msg = document.createElement("div");

        msg.id = "mensagemErro";

        msg.style.background = "#ffe5e5";
        msg.style.color = "#b00020";
        msg.style.border = "1px solid #ffb0b0";
        msg.style.padding = "10px";
        msg.style.marginBottom = "15px";
        msg.style.borderRadius = "6px";
        msg.style.display = "none";

        formEmpresa.prepend(msg);

    }

    return msg;

}

function mostrarMensagem(texto) {

    const msg = obterMensagem();

    msg.innerText = texto;

    msg.style.display = "block";

}

function limparMensagem() {

    const msg = document.getElementById("mensagemErro");

    if (msg) {

        msg.style.display = "none";

        msg.innerText = "";

    }

}


// =========================
// ENVIO AJAX
// =========================

formEmpresa.addEventListener("submit", async function (e) {

    // Somente para NOVO cadastro
    if (!formEmpresa.action.endsWith("/empresas/nova")) {

        return;

    }

    e.preventDefault();

    limparMensagem();

    const dados = new FormData(formEmpresa);

    try {

        const resposta = await fetch(formEmpresa.action, {

            method: "POST",

            body: dados

        });

        const json = await resposta.json();

        if (json.sucesso) {

            fecharModal();

            location.reload();

            return;

        }

        mostrarMensagem(json.mensagem);

    }

    catch (erro) {

        mostrarMensagem("Erro ao comunicar com o servidor.");

        console.error(erro);

    }

});
// =========================
// PESQUISA
// =========================

const pesquisa = document.getElementById("pesquisa");

if (pesquisa) {

    pesquisa.addEventListener("keyup", function () {

        let filtro = this.value.toLowerCase();

        let linhas = document.querySelectorAll("#tabelaEmpresas tbody tr");

        linhas.forEach(function (linha) {

            let texto = linha.innerText.toLowerCase();

            if (texto.indexOf(filtro) > -1) {

                linha.style.display = "";

            } else {

                linha.style.display = "none";

            }

        });

    });

}


// =========================
// EDITAR EMPRESA
// =========================

document.querySelectorAll(".editar").forEach(function (botao) {

    botao.addEventListener("click", function (e) {

        e.preventDefault();

        abrirModal();

        const id = this.dataset.id;

        document.getElementById("empresa_id").value = id;

        formEmpresa.action = "/empresas/editar/" + id;

        document.querySelector("[name=razao_social]").value =
            this.dataset.razao;

        document.querySelector("[name=nome_fantasia]").value =
            this.dataset.fantasia;

        let cnpj = this.dataset.cnpj.replace(/\D/g, "");

        if (cnpj.length === 14) {

            cnpj = cnpj.replace(/^(\d{2})(\d)/, "$1.$2");
            cnpj = cnpj.replace(/^(\d{2})\.(\d{3})(\d)/, "$1.$2.$3");
            cnpj = cnpj.replace(/\.(\d{3})(\d)/, ".$1/$2");
            cnpj = cnpj.replace(/(\d{4})(\d)/, "$1-$2");

        }

        document.querySelector("[name=cnpj]").value = cnpj;

        document.querySelector("[name=inscricao_estadual]").value =
            this.dataset.ie;

        document.querySelector("[name=responsavel]").value =
            this.dataset.responsavel;

        document.querySelector("[name=telefone]").value =
            this.dataset.telefone;

        document.querySelector("[name=whatsapp]").value =
            this.dataset.whatsapp;

        document.querySelector("[name=email]").value =
            this.dataset.email;

        document.querySelector("[name=cep]").value =
            this.dataset.cep;

        document.querySelector("[name=endereco]").value =
            this.dataset.endereco;

        document.querySelector("[name=numero]").value =
            this.dataset.numero;

        document.querySelector("[name=bairro]").value =
            this.dataset.bairro;

        document.querySelector("[name=cidade]").value =
            this.dataset.cidade;

        document.querySelector("[name=estado]").value =
            this.dataset.estado;

        document.querySelector("[name=status]").value =
            this.dataset.status;

        limparMensagem();

    });

});


// =========================
// SUBMIT DA EDIÇÃO
// =========================

formEmpresa.addEventListener("submit", function (e) {

    if (formEmpresa.action.endsWith("/empresas/nova")) {

        return;

    }

    // edição continua utilizando POST normal
    // exatamente como está no backend

});
// =========================
// MÁSCARA CNPJ
// =========================

function mascaraCNPJ(campo) {

    let valor = campo.value.replace(/\D/g, "");

    valor = valor.replace(/^(\d{2})(\d)/, "$1.$2");
    valor = valor.replace(/^(\d{2})\.(\d{3})(\d)/, "$1.$2.$3");
    valor = valor.replace(/\.(\d{3})(\d)/, ".$1/$2");
    valor = valor.replace(/(\d{4})(\d)/, "$1-$2");

    campo.value = valor.substring(0, 18);

}

document.querySelectorAll("[name=cnpj]").forEach(function (campo) {

    campo.addEventListener("input", function () {

        mascaraCNPJ(this);

    });

});


// =========================
// FORMATA CNPJ DA TABELA
// =========================

document.querySelectorAll(".cnpj").forEach(function (campo) {

    let valor = campo.innerText.replace(/\D/g, "");

    if (valor.length === 14) {

        valor = valor.replace(/^(\d{2})(\d)/, "$1.$2");
        valor = valor.replace(/^(\d{2})\.(\d{3})(\d)/, "$1.$2.$3");
        valor = valor.replace(/\.(\d{3})(\d)/, ".$1/$2");
        valor = valor.replace(/(\d{4})(\d)/, "$1-$2");

        campo.innerText = valor;

    }

});


// =========================
// FECHAR MODAL AO CLICAR FORA
// =========================

window.addEventListener("click", function (e) {

    if (e.target === modalEmpresa) {

        fecharModal();

    }

    if (e.target === modalExcluir) {

        fecharModalExcluir();

    }

});


// =========================
// TECLA ESC
// =========================

document.addEventListener("keydown", function (e) {

    if (e.key === "Escape") {

        fecharModal();

        fecharModalExcluir();

    }

});


// =========================
// FIM DO ARQUIVO
// =========================